packages/firestore/src/remote/rest_connection.ts

```diff
@@ -0,0 +1,168 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Token } from '../api/credentials';
+import { DatabaseId, DatabaseInfo } from '../core/database_info';
+import { SDK_VERSION } from '../../src/core/version';
+import { Connection, Stream } from './connection';
+import { logDebug, logWarn } from '../util/log';
+import { FirestoreError } from '../util/error';
+import { StringMap } from '../util/types';
+import { debugAssert } from '../util/assert';
+import { Indexable } from '../util/misc';
+
+const LOG_TAG = 'RestConnection';
+
+/**
+ * Maps RPC names to the corresponding REST endpoint name.
+ *
+ * We use array notation to avoid mangling.
+ */
+const RPC_NAME_URL_MAPPING: StringMap = {};
+
+RPC_NAME_URL_MAPPING['BatchGetDocuments'] = 'batchGet';
+RPC_NAME_URL_MAPPING['Commit'] = 'commit';
+RPC_NAME_URL_MAPPING['RunQuery'] = 'runQuery';
+
+const RPC_URL_VERSION = 'v1';
+const X_GOOG_API_CLIENT_VALUE = 'gl-js/ fire/' + SDK_VERSION;
+
+/**
+ * Base class for all Rest-based connections to the backend (WebChannel and
+ * HTTP).
+ */
+export abstract class RestConnection implements Connection {
```

Comment:
Within any given build isn't the case that there will ever be one connection type? It seems like the bundler could pick which "instance" you get and we could make these all free functions. Or is that coming in a separate PR?

Reply:
We can do this either way, but I thought it would be cleaner to use inheritance here. There will only ever be one implementation for this class in any given bundle.